homeassistant/components/recorder/table_managers/statistics_meta.py

```diff
@@ -228,23 +228,32 @@ def get_many(
                 "Providing statistic_type and statistic_source is mutually exclusive of statistic_ids"
             )
 
-        results: dict[str, tuple[int, StatisticMetaData]] = {}
-        missing_statistic_id: list[str] = []
-
-        for statistic_id in statistic_ids:
-            if id_meta := self._stat_id_to_id_meta.get(statistic_id):
-                results[statistic_id] = id_meta
-            else:
-                missing_statistic_id.append(statistic_id)
-
-        if not missing_statistic_id:
+        results = self.get_from_cache_threadsafe(statistic_ids)
+        if not (missing_statistic_id := statistic_ids.difference(results)):
             return results
 
         # Fetch metadata from the database
         return results | self._get_from_database(
             session, statistic_ids=missing_statistic_id
         )
 
+    def get_from_cache_threadsafe(
+        self, statistic_ids: set[str]
+    ) -> dict[str, tuple[int, StatisticMetaData]]:
+        """Get metadata from cache.
+
+        This call is thread safe and can be run in the event loop,
+        the database executor, or the recorder thread.
+        """
+        return {
+            statistic_id: meta
+            for statistic_id in statistic_ids
+            # We must use a get call here and never iterate over the dict
+            # because the dict can be modified by the recorder thread
+            # while we are iterating over it.
+            if (meta := self._stat_id_to_id_meta.get(statistic_id))
```

Comment:
```suggestion
            statistic_id: id_meta
            for statistic_id in statistic_ids
            # We must use a get call here and never iterate over the dict
            # because the dict can be modified by the recorder thread
            # while we are iterating over it.
            if (id_meta := self._stat_id_to_id_meta.get(statistic_id))
```